### ISSUE TITLE
Figure out what to do about external IPs

### ISSUE THREAD

**Issue by tallclair** [kubernetes/kubernetes#97110]
For CVE-2020-8554, https://github.com/kubernetes/kubernetes/issues/97076, we decided we couldn't patch it in-tree, and instead provided workarounds to disable (or allowlist) the feature through admission controls. Now that the issue is public, I'd like to open the conversation about a long-term fix.

Unless we missed something (entirely possible), I see a few possible paths forward:

1. Decide to stop supporting external IPs, deprecate and eventually remove the feature
2. Attempt to redesign the external IP feature, deprecate the old behavior, and manage the migration to the new feature
3. Accept the current state, and promote the [externalip-webhook](https://github.com/kubernetes-sigs/externalip-webhook) to a built-in admission controller
4. Accept the current state, and support the [externalip-webhook](https://github.com/kubernetes-sigs/externalip-webhook) as a long-term solution

/sig network architecture
/area security
/priority important-soon

**Comment by PushkarJ** [comment]
@tallclair : Thanks for sharing the details. Commenting here, since the original issue is closed. This maybe a novice question, but does default deny CNI network policy and / or Mutual Auth (e.g. mTLS) act as a mitigating control?

**Comment by chadswen** [comment]
Does this have the same root cause as the issue discussed in #22650 and fixed in openshift with an admission controller? (openshift issue: https://github.com/openshift/origin/issues/7808 and PR: https://github.com/openshift/origin/pull/7810) There is some existing discussion in the k/k thread if the added context helps form a strategy here.

I first encountered the issue a couple years ago when a tenant accidentally used an external IP incorrectly in one of our clusters, which I discovered while debugging an unreachable etcd node. The external IP Service was specifying the same IP as an etcd node's host IP, and that prevented all nodes running kube-proxy from reaching the etcd node since traffic was routed to the Service instead. Luckily we had other etcd nodes in the cluster and by debugging kube-proxy logs along with `ipvsadm` it didn't take too long to find the external IP as the root cause and shut it down.

I tried to report the issue via HackerOne several months ago, but I was never able to get the steps to reproduce working consistently in all clusters to meet the HackerOne report requirements, which may be similar to @champtar's experience in [this comment](https://github.com/kubernetes/kubernetes/issues/97076#issuecomment-740109980). The OPA Gatekeeper based solution is similar to the workaround I came to as well, but it would be great to fix it upstream.

I'm sure there are users that depend on external IP, but for what it's worth I have rarely seen it used correctly.

**Comment by tallclair** [comment]
Thanks for the additional context @chadswen, this does in fact look like a duplicate of https://github.com/kubernetes/kubernetes/issues/22650. Since the conversation on that issue didn't go anywhere, hopefully the added visibility of having a CVE associated will help drive progress.

**Comment by champtar** [comment]
@chadswen thanks for digging those old tickets, now I'm not sure I even searched for related open issues before reporting ...
Also I have a similar story, crashed a cluster to find this bug :)

Here is my write up with 18 tests on 3 clusters: https://blog.champtar.fr/K8S_MITM_LoadBalancer_ExternalIPs/
In those tests (that are a bit old now) kube-proxy IPVS is the most affected, and patching the LoadBalancer is the most reliable.

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/issues/97076 as a proposal for k/k

**Comment by champtar** [comment]
I never understood why the data plane is different between LoadBalancer and ExternalIPs.
That being said, you can completely remove ExternalIP and replace it by a small static LoadBalancer controller with a simple mapping as config `namespace/servicename -> ip`, or have people run MetalLB controller without MetalLB speaker.

**Comment by thockin** [comment]
> you can completely remove ExternalIP and replace it by a small static
LoadBalancer controller

A user could.  We can't unilaterally rip out a feature.  We don't want to
pull the rug out from underneath the (relatively small) number of
legitimate users.

**Comment by champtar** [comment]
I mean make it configurable, then at some point disable it by default, then even later remove it.

**Comment by thockin** [comment]
I'm proposing to give users a choice - completely off, completely on
(default, legacy), or enabled subject to your own policy.  I haven't
seen enough use-cases with enough commonality to warrant trying to add
more standard APIs to control it.  Users are adopting more general
policy control, and I don't think we should swimp against that
current.

**Comment by danwinship** [comment]
> 3. Accept the current state, and promote the [externalip-webhook](https://github.com/kubernetes-sigs/externalip-webhook) to a built-in admission controller
> 4. Accept the current state, and support the [externalip-webhook](https://github.com/kubernetes-sigs/externalip-webhook) as a long-term solution

Note that externalip-webhook's `allowed-external-ip-cidrs` solves the problem from the CVE but is still insecure if there are mutually-untrusting users, because it doesn't stop one user from creating a Service using the same externalIP+port as another user's existing Service.

**Comment by thockin** [comment]
Right.  To *really* solve this we need a much more robust control API and I
am not (yet) convinced that is justified.

**Comment by tgraf** [comment]
As a potential reference on how to fix this: We have fixed this in Cilium a while ago [0] by ignoring ExternalIP for any traffic from a pod unless the destination IP is mapped to a node IP (GCE case). It appears that this code [1] is trying to do something similar but is specific to bridge based network implementations.

[0] https://github.com/cilium/cilium/pull/12220/commits/1d8589a9877e348055b3a0d4b30984377f8dcc32
[1] https://github.com/kubernetes/kubernetes/blob/018942a92a90f91261ba63971827f2ec6b64587e/pkg/proxy/ipvs/proxier.go#L1740-L1752

**Comment by thockin** [comment]
The problem is that the feature is too loosely defined.  Something like
that feels right but is, strictly, a breaking change *and* doesn't solve
the whole problem (i.e. I can still steal your IP address).

I just want to break it completely and let people who *need* it define
their own policies in their own ways...

**Comment by tgraf** [comment]
I'm not sure I understand how you can still steal the IP outside of what ExternalIP was intended for unless you are referring to a potential restriction for ExternalIPs to not cover PodIP and ServiceIP CIDR which would potentially be even better. I'm not sure what we can better at the k8s service implementation layer. I agree with your overall statement though. We often see users try and patch over this with network policies. Maybe it would be cleaner to have separate k8s resources for cluster internal and external services.

**Comment by champtar** [comment]
You can have multiple Services with the same  EXTERNAL-IP:PORT
```# kubectl get svc -n kubeproxy-mitm
NAME                    TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)        AGE
mitm-external-eip-dns   ClusterIP      10.233.55.182   8.8.8.8       53/UDP         4m10s
mitm-external-lb-dns    LoadBalancer   10.233.35.158   8.8.8.8       53:31303/UDP   64s

# kubectl get svc -n kubeproxy-mitm
NAME                    TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)        AGE
mitm-external-lb-dns1   LoadBalancer   10.233.47.145   8.8.8.8       53:31545/UDP   4m37s
mitm-external-lb-dns2   LoadBalancer   10.233.40.23    8.8.8.8       53:31556/UDP   4m37s
mitm-external-lb-dns3   LoadBalancer   10.233.28.107   8.8.8.8       53:31206/UDP   4m37s
```

**Comment by vassilvk** [comment]
@danwinship 

> Note that externalip-webhook's allowed-external-ip-cidrs solves the problem from the CVE but is still insecure if there are mutually-untrusting users, because it doesn't stop one user from creating a Service using the same externalIP+port as another user's existing Service.

As an alternative to using **external-webhook** one can use a [KubeMod](https://github.com/kubemod/kubemod) reject rule like [this one](https://github.com/kubemod/kubemod#resource-rejection).

KubeMod rules apply only to the namespace they are deployed to.
This makes it possible to have different policy rules with different allowed CIDRs and IP ranges for each tenant in a multi-tenant cluster.

**Comment by cloudsriseup** [comment]
I concur with @tgraf .  It would be much cleaner to separate k8s resources for internal / external cluster services.

**Comment by pmoust** [comment]
> It would be much cleaner to separate k8s resources for internal / external cluster services.

It's a good to have that classification.
But it doesn't solve for the intra-cluster MITM case, unless I am missing something.
Perhaps added controls (like namespace classification) would help.

**Comment by thockin** [comment]
Gateway should be the basis for the distinction between internal and external.

**Comment by thockin** [comment]
Proposd impl: https://github.com/kubernetes/kubernetes/pull/97395

**Comment by aojea** [comment]
should we close it now @thockin ?

**Comment by zhousam** [comment]
> Proposd impl: #97395

is this PR the final solution to CVE-2020-8554？Do we have plans to imple the second path "Attempt to redesign the external IP feature, deprecate the old behavior, and manage the migration to the new feature" in future version of kubernetes?

**Comment by cbron** [comment]
The question by @zhousam still needs to be addressed. If the PR is merged and already live in the now released k8s 1.21, should this issue be resolved or is there additional work needed ? cc @thockin @tallclair

**Comment by aojea** [comment]
> is this PR the final solution to CVE-2020-8554？Do we have plans to imple the second path "Attempt to redesign the external IP feature, deprecate the old behavior, and manage the migration to the new feature" in future version of kubernetes?

The field is part of Services API that is GA, is not likely to be removed, see comment https://github.com/kubernetes/kubernetes/issues/97110#issuecomment-740714200

> Gateway should be the basis for the distinction between internal and external.

@thockin refers to https://gateway-api.sigs.k8s.io/ , I think that he means that any attempt to redesign will come from that new API


Hope that answers the remaining questions
/close
